pkg/controller/actions/deploy/action_deploy.go

```diff
@@ -94,69 +99,130 @@ func WithAnnotations(values map[string]string) ActionOpts {
-		resources.SetAnnotations(&obj, a.annotations)
-
-		resources.SetAnnotation(&obj, annotations.ComponentGeneration, strconv.FormatInt(rr.Instance.GetGeneration(), 10))
-		resources.SetAnnotation(&obj, annotations.PlatformType, string(rr.Release.Name))
-		resources.SetAnnotation(&obj, annotations.PlatformVersion, rr.Release.Version.String())
-
-		switch {
-		// the user has explicitly marked the current object as not owned by the operator, so
-		// skip any further processing
-		case current != nil && resources.GetAnnotation(current, annotations.ManagedByODHOperator) == "false":
-			continue
-
-		// The object is explicitly marked as not owned by the operator in the manifests,
-		// so it should be created if it doesn't exist, but should not be modified afterward.
-		case resources.GetAnnotation(&obj, annotations.ManagedByODHOperator) == "false":
-			// remove the opendatahub.io/managed as it should not be set
-			// to the actual object in this case
-			resources.RemoveAnnotation(&obj, annotations.ManagedByODHOperator)
-
-			err := a.create(ctx, rr.Client, obj)
-			if err != nil {
-				return err
-			}
+	return nil
+}
 
-		default:
-			owned := rr.Manager.Owns(obj.GroupVersionKind())
-			if owned {
-				if err := ctrl.SetControllerReference(rr.Instance, &obj, rr.Client.Scheme()); err != nil {
-					return err
-				}
-			}
+func (a *Action) deploy(
+	ctx context.Context,
+	rr *odhTypes.ReconciliationRequest,
+	obj unstructured.Unstructured,
+) (bool, error) {
+	current, lookupErr := a.lookup(ctx, rr.Client, obj)
+	if lookupErr != nil && !k8serr.IsNotFound(lookupErr) {
+		return false, fmt.Errorf("failed to lookup object %s/%s: %w", obj.GetNamespace(), obj.GetName(), lookupErr)
+	}
 
-			var err error
+	resources.SetLabels(&obj, a.labels)
+	resources.SetAnnotations(&obj, a.annotations)
 
-			switch a.deployMode {
-			case ModePatch:
-				err = a.patch(ctx, rr.Client, obj, current)
-			case ModeSSA:
-				err = a.apply(ctx, rr.Client, obj, current)
-			default:
-				err = fmt.Errorf("unsupported deploy mode %s", a.deployMode)
-			}
+	resources.SetAnnotation(&obj, annotations.ComponentGeneration, strconv.FormatInt(rr.Instance.GetGeneration(), 10))
+	resources.SetAnnotation(&obj, annotations.PlatformType, string(rr.Release.Name))
+	resources.SetAnnotation(&obj, annotations.PlatformVersion, rr.Release.Version.String())
+
+	// backup copy for caching
+	origObj := obj.DeepCopy()
+
+	if a.cache != nil && current != nil {
+		ck, err := a.computeCacheKey(current, &obj)
+		if err != nil {
+			return false, fmt.Errorf("failed to compute request identifier: %w", err)
+		}
+
+		// no changes, no need to re-deploy it
+		if a.cache.Has(ck) {
+			return false, nil
+		}
+	}
+
+	var deployedObj *unstructured.Unstructured
+
+	switch {
+	// the user has explicitly marked the current object as not owned by the operator, so
+	// skip any further processing
+	case current != nil && resources.GetAnnotation(current, annotations.ManagedByODHOperator) == "false":
```

Comment:
It's probably not worth it if it complicates the code, and I don't know if it would save much time or space: I wonder if it would be worth handling this case earlier, around L139, which would save the deep copy in to `origObj`, computing the cache key, and cache lookup, for resources that fall into this case?

Reply:
it makes sense, thx!

Reply:
done